File: PA95.java
//Leetcode: 253. Meeting Rooms II
//Time Complexity: O(nlogn) Building heap is nlogn. where n is number of meeting intervals
//Space Complexity: O(n), Heap size will be at max equal to number of intervals
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        if(intervals==null || intervals.length==0) return 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((int[] a,int[] b)->(a[0]==b[0]?(a[1]-b[1]):(a[0]-b[0])));
        
        for(int[] in : intervals){
            pq.offer(in);
        }
        PriorityQueue<int[]> pqEnd= new PriorityQueue<>((int[] a,int[] b)->(a[1]-b[1]));
        pqEnd.offer(pq.poll());
        while(pq.size()>0){
            int[] meeting = pq.poll();
            int[] oldMeeting =pqEnd.peek();
            if(oldMeeting[1]>meeting[0]){
                
                pqEnd.offer(meeting);
            }
            else{
                oldMeeting= pqEnd.poll();
                oldMeeting[1]= meeting[1];
                pqEnd.offer(oldMeeting);
            }
        }
        
        return pqEnd.size() ;
        
    }
}

File: PA95.py
#Time Complexity - O(nlogn)
#Space Complexity - O(n)
import heapq
class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        if not intervals:
            return 0
        intervals = sorted(intervals, key = lambda x : x[0])
        heap = []
        heapq.heappush(heap,intervals[0][1])
        for i in range(1,len(intervals)):
            if intervals[i][0] >= heap[0]:
                heapq.heappop(heap)
            heapq.heappush(heap,intervals[i][1])
        
        return len(heap)